src/app/CommandHandler.h

```diff
@@ -473,6 +446,54 @@ class CommandHandler
         AwaitingDestruction, ///< The object has completed its work and is awaiting destruction by the application.
     };
 
+    /**
+     * @brief Best effort to add InvokeResponse to InvokeResponseMessage.
+     *
+     * Tries to add response using lambda. Upon failure to add response, attempts
+     * to rollback the InvokeResponseMessage to a known good state. If failure is due
+     * to insufficient space in the current InvokeResponseMessage:
+     *  - Finalizes the current InvokeResponseMessage.
+     *  - Allocates a new InvokeResponseMessage.
+     *  - Reattempts to add the InvokeResponse to the new InvokeResponseMessage.
+     *
+     * @param [in] addResponseFunction A lambda function responsible for adding the
+     *             response to the current InvokeResponseMessage.
+     */
+    template <typename Function>
+    CHIP_ERROR TryAddingResponse(Function && addResponseFunction)
+    {
+        // Invalidate any existing rollback backups. The addResponseFunction is
+        // expected to create a new backup during either PrepareInvokeResponseCommand
+        // or PrepareStatus execution. Direct invocation of
+        // CreateBackupForResponseRollback is avoided since the buffer used by
+        // InvokeResponseMessage might not be allocated until a Prepare* function
+        // is called.
+        mRollbackBackupValid = false;
+        CHIP_ERROR err       = addResponseFunction();
+        if (err == CHIP_NO_ERROR)
+        {
+            return CHIP_NO_ERROR;
+        }
+        ReturnErrorOnFailure(RollbackResponse());
+        // If we failed to add a command due to lack of space in the
+        // packet, we will make another attempt to add the response using
+        // an additional InvokeResponseMessage.
+        if (mState != State::AddedCommand || err != CHIP_ERROR_NO_MEMORY)
+        {
+            return err;
+        }
+        ReturnErrorOnFailure(FinalizeInvokeResponseMessageAndPrepareNext());
+        err = addResponseFunction();
+        if (err != CHIP_NO_ERROR)
+        {
+            // The return value of RollbackResponse is ignored, as we prioritize
+            // conveying the error generated by addResponseFunction to the
+            // caller.
+            RollbackResponse();
```

Comment:
is rollbackreponse logging the failure? otherwise it seems valuable to log here for debug purposes, otherwise the error is forever lost.